File: 98/cpp/solutions.cpp
#include "nodes.h"

#include "lib/trees.cpp"

using namespace nodes;
using namespace lib;

class Solution1
{
public:
    bool isValidBST(BinaryTreeNode *root)
    {
        BinarySearchTree bst(root);
        return bst.isValid();
    }
};

//inorder traversal.
//inorder Rule: left(all nodes) < root < right(all nodes), or prev < cur.
class Scratch1
{
public:
    bool isValidBST(BinaryTreeNode *root)
    {
        BinaryTreeNode *prev = nullptr;
        return isValid(root, prev);
    }

private:
    bool isValid(BinaryTreeNode *root, BinaryTreeNode *&prev)
    {
        // Terminator
        if (!root)
            return true;

        // Left Tree. Recursive case
        if (!isValid(root->left, prev))
            return false;

        // Root. Process or Logic, Action.
        if (prev && prev->val >= root->val)
            return false;

        // Right Tree. Recursive case
        prev = root;
        return isValid(root->right, prev);
    }
};

File: 98/python/isValidBST/solutions.py
from lib.trees import BinarySearchTree
from nodes import BinaryTreeNode


class Solution1:
    def isValidBST(self, root: BinaryTreeNode) -> bool:
        return BinarySearchTree(root).is_valid()


# inorder traversal
# inorder Rule: left(all nodes) < root < right(all nodes), or prev < cur.
class Scratch1:
    def isValidBST(self, root: BinaryTreeNode) -> bool:
        self.prev = None
        return self._is_valid(root)

    def _is_valid(self, root: BinaryTreeNode) -> bool:
        # Terminator
        if not root: return True

        # Left Tree
        if not self._is_valid(root.left):
            return False

        # Process or Logic
        if self.prev and self.prev.val >= root.val:
            return False

        # Right Tree, Recursive case
        self.prev = root
        return self._is_valid(root.right)
